src/libraries/Microsoft.Extensions.Hosting/src/HostApplicationBuilder.cs

```diff
@@ -0,0 +1,435 @@
+                if (_serviceProviderFactory is not null)
+                {
+                    _hostApplicationBuilder._createServiceProvider = () =>
+                    {
+                        object containerBuilder = _serviceProviderFactory.CreateBuilder(_hostApplicationBuilder.Services);
+                        _hostApplicationBuilder._configureContainer(containerBuilder);
+                        return _serviceProviderFactory.CreateServiceProvider(containerBuilder);
+                    };
+                }
+            }
+
+            public IDictionary<object, object> Properties => _hostApplicationBuilder._hostBuilderContext.Properties;
+
+            public IHost Build() => throw new NotSupportedException();
+
+            public IHostBuilder ConfigureHostConfiguration(Action<IConfigurationBuilder> configureDelegate)
+            {
+                _configureHostConfigActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+                return this;
+            }
+
+            public IHostBuilder ConfigureAppConfiguration(Action<HostBuilderContext, IConfigurationBuilder> configureDelegate)
+            {
+                _configureAppConfigActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+                return this;
+            }
+
+            public IHostBuilder ConfigureServices(Action<HostBuilderContext, IServiceCollection> configureDelegate)
+            {
+                _configureServicesActions.Add(configureDelegate ?? throw new ArgumentNullException(nameof(configureDelegate)));
+                return this;
+            }
+
+            public IHostBuilder UseServiceProviderFactory<TContainerBuilder>(IServiceProviderFactory<TContainerBuilder> factory) where TContainerBuilder : notnull
+            {
+                _serviceProviderFactory = new ServiceFactoryAdapter<TContainerBuilder>(factory ?? throw new ArgumentNullException(nameof(factory)));
+                return this;
+
+            }
+
+            public IHostBuilder UseServiceProviderFactory<TContainerBuilder>(Func<HostBuilderContext, IServiceProviderFactory<TContainerBuilder>> factory) where TContainerBuilder : notnull
+            {
+                _serviceProviderFactory = new ServiceFactoryAdapter<TContainerBuilder>(() => _hostApplicationBuilder._hostBuilderContext, factory ?? throw new ArgumentNullException(nameof(factory)));
+                return this;
+            }
+
+            public IHostBuilder ConfigureContainer<TContainerBuilder>(Action<HostBuilderContext, TContainerBuilder> configureDelegate)
+            {
+                _configureContainerActions.Add(new ConfigureContainerAdapter<TContainerBuilder>(configureDelegate
+                    ?? throw new ArgumentNullException(nameof(configureDelegate))));
+
+                return this;
+            }
+        }
+
+        internal sealed class CheckedServiceCollection : IServiceCollection
```

Comment:
I like this idea. It would require adding new public API to ServiceCollection to trigger making it readonly. What do you think about an instance method call `MakeReadOnly()`? I don't like adding an `IsReadonly` setter as public API because that would allow transitioning back to not being readonly. If this makes sense I can file a new API proposal. I don't think it should block this PR though.

Reply:
> What do you think about an instance method call MakeReadOnly()? 

I like that.

>  If this makes sense I can file a new API proposal. 

Yes, please. If you have time.

> I don't think it should block this PR though.

Agreed.